src/main/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverter.java

```diff
@@ -205,6 +197,13 @@ public void afterPropertiesSet() {
 					targetEntity.getPropertyAccessor(result).setProperty(versionProperty, version);
 				}
 			}
+
+			if (targetEntity.hasSeqNoPrimaryTermProperty() && document.hasSeqNo() && document.hasPrimaryTerm()) {
+				SeqNoPrimaryTerm.ofAssigned(document.getSeqNo(), document.getPrimaryTerm()).ifPresent(seqNoPrimaryTerm -> {
+					ElasticsearchPersistentProperty property = targetEntity.getRequiredSeqNoPrimaryTermProperty();
+					targetEntity.getPropertyAccessor(result).setProperty(property, seqNoPrimaryTerm);
+				});
+			}
```

Comment:
```suggestion
			if (targetEntity.hasSeqNoPrimaryTermProperty() && document.hasSeqNo() && document.hasPrimaryTerm()) {
				SeqNoPrimaryTerm seqNoPrimaryTerm = new SeqNoPrimaryTerm(document.getSeqNo(), document.getPrimaryTerm());
					ElasticsearchPersistentProperty property = targetEntity.getRequiredSeqNoPrimaryTermProperty();
					targetEntity.getPropertyAccessor(result).setProperty(property, seqNoPrimaryTerm);
			}
```
if the `SeqNoPrimaryTerm` constructor is added as I wrote there

Reply:
Elasticsearch can actually return 'unassigned' values for seq_no and primary_term. Earlier you suggested to return them 'as is', just as numbers, without any interpretation, from `Document.getSeqNo()` and `Document.getPrimaryTerm()`, so the analysis of their 'validity' was moved to the 'optional factory' method, namely `SeqNoPrimaryTerm.ofAssigned()`. But if we remove that factory method and just invoke a constructor that blows up on an 'unassigned' value, it will throw an exception in such cases.

I suggest to do the following:
1. Make the constructor public to allow Jackson use it for instance construction
2. Still leave `ofAssigned()` and use it here (in the mapping code) to sort out the cases when Elasticsearch returns unassigned values
3. Add assertions to the constructor to avoid garbage going from some kind of a REST API.

What do you think?

Reply:
Where is this documented? Then I'd suggest in `MappingElasticsearchConverter:202` to check the values of `document.getSeqNo()` and `document.getPrimaryTerm()` with a simple `if` check and set the `SeqNoPrimaryTerm` if these are correct. This is the only place where this check is used.

As https://www.elastic.co/guide/en/elasticsearch/reference/current/optimistic-concurrency-control.html does not talk about potentially unassigned values I  think it might be confusing to introduce this into the code of `SeqNoPrimaryTerm`.

We still need the `IllegalArgumentException in case of the user setting some bad values.

Reply:
The `search` API does not return `seq_no` and `primary_term` by default, you need to specifically ask it to return the real values (using `seq_no_primary_term` field), otherwise it returns 'unassigned' values. We send `seq_no_primary_term=true` when doing searches, but it seems a bit fragile to hope that unassigned values will never be returned by Elasticserch. Maybe a user will manually send a request without `seq_no_primary_term` and will want to use our mapping code to map, or we've omitted some API that needs this to be specifically enabled.